compiler/modules/front/codeGen.bal

```diff
@@ -1939,39 +1974,63 @@ final readonly & bir:UniformOrderType[] UNIFORM_ORDER_TYPES = [t:UT_BOOLEAN, t:U
 function operandOrderType(CodeGenContext cx, bir:Operand operand) returns bir:OrderType? {
     if operand is bir:Register {
         t:SemType operandTy = operand.semType;
-        if operandTy === t:NIL {
-            return ();
-        }
-        t:UniformTypeBitSet arrTy = t:LIST;
-        if t:isSubtypeSimple(operandTy, arrTy) {
-            t:UniformTypeBitSet? memberTy = t:simpleArrayMemberType(cx.mod.tc, operandTy);
-            if memberTy is t:UniformTypeBitSet {
-                bir:OrderType? ot = operandUniformOrderType(memberTy);
-                if ot is bir:UniformOrderType {
-                    return <bir:ArrayOrderType> [{opt:ot}];
-                }
-                if ot is bir:OptOrderType {
-                    return <bir:ArrayOrderType> [ot];
-                }
-            }
-            else {
-                panic err:impossible("Failed to get array member type");
-            }
-        }
-        else {
-            return operandUniformOrderType(operandTy);
-        }
+        return toOrderType(cx, operandTy);
     }
     else {
         var tc = t:constUniformTypeCode(operand);
-        if tc is bir:UniformOrderType {
+        if tc is t:UT_NIL {
+            return bir:nilOrderType;
+        }
+        else if tc is bir:UniformOrderType {
             return tc;
         }
     }
     return ();
 }
 
+function toOrderType(CodeGenContext cx, t:SemType operandTy) returns bir:OrderType? {
+    OrderTypeMemo? memoized = cx.orderTyMemo[operandTy];
+    if memoized != () {
+        return memoized.orderType;
+    }
+    OrderTypeMemo memo = { semType: operandTy };
+    // XXX We assume recursive types can't be OrderType, reconsider after adding tuples.
+    cx.orderTyMemo.add(memo);
+
+    bir:OrderType? result;
+    if operandTy == t:NIL {
+        result = bir:nilOrderType;
+    }
+    t:UniformTypeBitSet arrTy = t:LIST;
+    if t:isSubtypeSimple(operandTy, arrTy) {
+        t:SemType? memberTy = t:listMemberType(cx.mod.tc, operandTy);
+        if memberTy == t:NEVER {
+            result = bir:emptyTupleOrderType;
+        }
+        else if memberTy is t:SemType {
+            bir:OrderType? memberOrderType = toOrderType(cx, memberTy);
+            result = memberOrderType == () ? () : { memberOrderType };
+        }
+        else {
+            result = ();
+        }
+    }
+    else if t:containsNil(operandTy) {
+        // Cast is valid since we exclude NIL resulting non-opt order type.
+        var opt = <bir:UniformOrderType|bir:ArrayOrderType|()> toOrderType(cx, t:diff(operandTy, t:NIL));
+        result = opt == () ? () : { opt };
+    }
+    else {
+        result = operandUniformOrderType(operandTy);
+    }
+    memo.orderType = result;
+    return result;
+}
+
 function operandUniformOrderType(t:SemType operandType) returns bir:OrderType? {
```

Comment:
Use Basic rather than Uniform here to go with suggested change in bir.

Then call this semTypeBasicOrderType to go with suggested name change in toOrderType.